src/aklite_client_ext.cc

```diff
@@ -0,0 +1,146 @@
+
+#include "http/httpclient.h"
+#include "libaktualizr/config.h"
+#include "liteclient.h"
+#include "primary/reportqueue.h"
+
+#include "aktualizr-lite/tuf/tuf.h"
+#include "composeapp/appengine.h"
+#include "composeappmanager.h"
+#include "docker/restorableappengine.h"
+#include "ostree/repo.h"
+#include "tuf/akhttpsreposource.h"
+#include "tuf/akrepo.h"
+#include "tuf/localreposource.h"
+#include "uptane/exceptions.h"
+
+// TODO: Reuse this
+static Uptane::Target toUptaneTarget(const TufTarget& t) {
+  Json::Value target_json;
+  target_json["hashes"]["sha256"] = t.Sha256Hash();
+  target_json["custom"]["targetFormat"] = "OSTREE";
+  target_json["length"] = 0;
+  Uptane::Target target(t.Name(), target_json);
+  return target;
+}
+
+// Returns the target that should be installed, if any.
+// It might be an updated version, a rollback target, or even the currently installed target, in case we need to sync
+// apps
+CheckInResult AkliteClientExt::GetTargetToInstall(const LocalUpdateSource* local_update_source, int version,
+                                                  const std::string& target_name, const TufTarget& current,
+                                                  bool allow_bad_target, bool force_apps_sync) {
+  // Tell CheckIn and CheckInLocal methods to not call notifyTufUpdateFinished in case of success
+  usingUpdateClientApi = true;
+  client_->setAppsNotChecked();
+  bool rollback_operation = false;
+
+  CheckInResult res = local_update_source == nullptr ? CheckIn() : CheckInLocal(local_update_source);
+
+  if (res.status != CheckInResult::Status::Ok && res.status != CheckInResult::Status::OkCached) {
+    LOG_WARNING << "Unable to update latest metadata";
+    return res;
+  }
+
+  auto candidate_target = res.SelectTarget(version, target_name);
+  if (candidate_target.IsUnknown()) {
+    std::string err = "No matching target";
+    client_->notifyTufUpdateFinished(err);
+    LOG_WARNING << err;
+    res.status = CheckInResult::Status::TufTargetNotFound;
+    return res;
+  }
+
+  if (IsRollback(current) && current.Name() == candidate_target.Name()) {
+    // Handle the case when Apps failed to start on boot just after an update.
+    // This is only possible with `pacman.create_containers_before_reboot = 0`.
+    LOG_INFO << "The currently booted Target is a failing Target, finding Target to rollback to...";
+    const TufTarget rollback_target = Target::toTufTarget(client_->getRollbackTarget());
+    if (rollback_target.IsUnknown()) {
+      const auto err{boost::str(boost::format("Failed to find Target to rollback to after a failure to start "
+                                              "Apps at boot on a new version of sysroot;"
+                                              " failing current Target: %s, hash: %s") %
+                                current.Name() % current.Sha256Hash())};
+
+      LOG_ERROR << err;
+      res.status = CheckInResult::Status::Failed;  // TODO: Specify proper error status
```

Comment:
It looks like we need to extend the checkinresult type, and use another return type for `GetTargetToInstall`.

Reply:
This specific situation was only being handled in the aklite daemon. Should we add a new status now (and documentation) for it to be identifiable for cli users? Something like `NoTargetToRollbackTo`?

Reply:
Code was adjusted to return `RollbackTargetNotFound` in that case.